### ISSUE TITLE
A path toward supporting writeable views in mygrad

### ISSUE THREAD

**Issue by rsokl** [rsokl/MyGrad#296]
@aslvrstn @davidmascharka @petarmhg @Zac-HD 
(For people tagged in this issue, I would appreciate to hear your thoughts on this, if you have any)

Currently there is no mechanism in mygrad to writing in-place to a view to affect a base tensor. E.g. the following behavior is desirable

```python
# we can't do this yet
>>> a = mg.zeros((3, 3), int);
>> mg.fill_diagonal(mg.fliplr(a), [1, 2, 3])  # Horizontal flip and write to its diag
>>> a
Tensor([[0, 0, 1],
        [0, 2, 0],
        [3, 0, 0]])
```

as of now `mg.fill_diagonal` - as an in-place operation - would first make a copy of the output of `mg.fliplr(a)` before writing to its diagonal, effectively rendering ` mg.fill_diagonal(mg.fliplr(a), [1,2,3])` a no-op. Accommodating the fact that the in-place operation is acting on a view of a tensor opens a whole can of worms...

#295 are my baby steps towards tracking and enabling writeable tensor views in mygrad. Towards this end mygrad needs to add:
- mechanisms for distinguishing between ops which produce distinct tensors from those that produce views
- the ability to rewrite the history of a base-tensor, which is being updated in-place, that participates in both view-ops and standard ops

Eventually there will also be the need to restrict array-write ability such that arrays involved in an "active" computational graph are not mutated outside of the computational graph (although there is no way to guarantee this)

### A substantial outstanding issue

Once #295 is merged, there is a major outstanding issue that must be addressed, which is that `Tensor.__init__` currently does not make copies of tensor data. Thus a tensor created from another will effectively be a view of it, yet this this not tracked. E.g.

```python
>>> x = mg.arange(10)
>>> y = Tensor(x)
>>> y.base is None  # `y` doesn't look like a view of `x`
True
>>> y.data is x.data # yet they point to the exact same data
True
```

in this instance a mutation of `y` will mutate `x` without being tracked by `x`'s computational graph. A bizarre way of patching this up is to treat `y = Tensor(x)` like `y = x[...]`, but this is bound to be confounding for users who would expect `y` and `x` to be distinct.

Thus my plan is the following:

- [x] Begin tracking memory ownership - as is done in numpy - via `tensor.base` (#295)
- [x] Simplify the process of managing the computational graph by removing the process of having users manually null-gradients and clear the computational graph. Instead, the graph is cleared during backprop (#301)
- [x] Any array that participates in an "active" computational graph will be set to read-only (this will be lifted once backprop is completed)
- [x] The memory locking behavior is annoying / bad if you just want use MyGrad to do an evaluation. Will introduce an "inference-mode" where no graph-tracking or memory-locking is performed. A context manager and decorator will be provided to manage this.
- [x] Only in-place operations performed by MyGrad will be able to (temporarily) write to arrays in-place
- [x] `Tensor(x, constant=False)` will copy `x`, if the tensor is created by a user
- [x] `Tensor(x, constant=True)` will not copy `x`, but in-place operations on constant tensors will also be forbidden
- [x] Any base tensor – one which serves as views of other tensors – will have access to those views, so that it can do surgery on relevant parts of the computational graph when it is involved in an in-place operation

Why copy data when creating variable tensors? Consider the following:

```python
>>> arr = np.arrange(3)
>>> x = Tensor(arr)
>>> y = Tensor(arr)  # there is no feasible way for mygrad to see that `arr` has been made a tensor elsewhere
>>> x *= 2  # arr is temporarily made writeable to facilitate this
>>> y  # mutated but this change is not tracked by the computational graph
Tensor([0, 2, 6]) 
```

without any mechanism for logging or marking numpy arrays that have been made tensors elsewhere, there doesn't seem to be a way to "responsibly" permit mutations in the computational graph.

Why make constant tensors immutable - why not treat them like variable tensors? Because constant tensors are created on the fly all of the time - any time mygrad operation encounters a non-tensor object it gets cast to a constant tensor, and I am concerned about the overhead of making copies of all of these things.

The most dramatic case of this is when training a DNN:

```python
for batch_ndarray in data:
    model(batch_ndarray )  # batch_ndarray gets cast to constant-tensor on each forward pass
```

whereas the model's weights need only be initialized - and thus copied - once, the batches of data will be copied repeatedly if we were to permit mutable constant tensors.

**Comment by rsokl** [comment]
This is done!
